### PR TITLE
allow clients to set QUERY_TAG parameter via context

### PR DESCRIPTION
**Pull request by GregOwen**
### Description
1. Add new `WithQueryTag` method to allow clients to pass the `QUERY_TAG` parameter to a query via the context
2. Add test to verify that we're setting the tag correctly

### Checklist
- [X] Code compiles correctly
- [ ] Run ``make fmt`` to fix inconsistent formats
- [X] Run ``make lint`` to get lint errors and fix all of them
- [X] Created tests which fail without the change (if possible)
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by GregOwen** [comment]
@sfc-gh-jbahk here's the second of the two PRs I mentioned. This one should hopefully be pretty straightforward.

**Comment by GregOwen** [comment]
@sfc-gh-jbahk Sorry for the delay in getting back to you. Some of our customers have asked us to include Sigma-level metadata (username, Sigma document) for each Snowflake query so that they can do automated attribution of their Sigma-initiated Snowflake usage. Right now, we add this metadata as a JSON blob in a comment at the end of the query, but our customers have asked us to move this information into the `Query Tag` column so that it's easier for them to extract and analyze within Snowflake.

**Comment by sfc-gh-jbahk** [comment]
@GregOwen I see potential conflicts with this approach wherein the statement parameters that the customers might pass into the statement could either throw errors, not accepted by our server or conflict with our internal parameters. Another issue would be security threats including, but not limited to, injections. Could you give me a more specific example? Thanks.

**Comment by GregOwen** [comment]
That makes sense. We probably don't need this to be as wide an interface as it currently is.
Our requirements are:
1. we can pass a JSON blob containing metadata per-query
2. the per-query metadata will be associated with the query in a way that Sigma customers can parse and analyze within Snowflake (in particular, we don't want the JSON blob to be part of a comment within the text of the query itself)

The way that we're currently using it is something like this:
```
params := make(map[string]string)
params["QUERY_TAG"] = fmt.Sprintf(`{"request-id":"%s","email":"%s"}`, requestId, email) // possibly Sigma document ID, etc.
ctx = WithContextStatementParameters(ctx, params)

// execute the query with the given context
```

**Comment by sfc-gh-jbahk** [comment]
@GregOwen I see. Where would these parameters be used? Will it be inserted as a value of a column?

**Comment by GregOwen** [comment]
@sfc-gh-jbahk The idea is that the metadata will show up as a `Query Tag` column in the customer's History view. Here's an example:

Screen Shot 2021-05-13 at 1 50 22 PM

This way a customer's Snowflake admin can use Snowflake to ask questions like "how many queries did Sigma run?" or "which Sigma documents are responsible for my increased Snowflake usage?".

**Comment by sfc-gh-jbahk** [comment]
@GregOwen I've reviewed this with my team and think it should be okay to move forward but if you can change the code to be less generic such that the statement context doesn't accept anything the user inputs but rather tailor it towards a query tag. Maybe this can be of some use: https://docs.snowflake.com/en/sql-reference/parameters.html#query-tag.

**Comment by GregOwen** [comment]
@sfc-gh-jbahk sounds great - I reworked it to be more precisely targeted at `QUERY_TAG` and added a test

**Comment by sfc-gh-jbahk** [comment]
@GregOwen can you rebase and push this branch?

**Comment by GregOwen** [comment]
@sfc-gh-jbahk done

**Comment by GregOwen** [comment]
Test failures look like they're related to test infrastructure/secrets:

```
Run ./ci/test.sh
  ./ci/test.sh
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.15.13/x64
    PARAMETERS_SECRET: 
    CLOUD_PROVIDER: AWS
gpg: gcry_kdf_derive failed: Invalid data
gpg: decryption failed: No secret key
Error: Process completed with exit code 2.
```

**Comment by tampajohn** [comment]
@sfc-gh-jbahk any idea if/when this is planned to be merged?  Or what needs to be done to get it mergeable?

**Comment by sfc-gh-jbahk** [comment]
@sfc-gh-cshi can you take a look at this PR and see if it's something the go driver needs/wants? cc: @sfc-gh-hkapre

**Comment by DmitryMaletin** [comment]
@sfc-gh-jbahk  is there any plan to complete the merge?

**Comment by sfc-gh-pfus** [comment]
Hi @tampajohn ! Do you still need this? Can you rebase and solve conflicts and align to new test helper function?

**Comment by sfc-gh-pfus** [comment]
Sorry @tampajohn , I meant @GregOwen of course!

**Comment by GregOwen** [comment]
@sfc-gh-pfus yes, we would still like to merge this PR. Can you confirm that somebody at Snowflake will be responsive to changes to this PR? Last time we got it to the 1-yard line but couldn't get it merged because each new commit requires manual re-approval to run the CI tests.

**Comment by sfc-gh-pfus** [comment]
Yes, now go driver migrated to another team. I merged/closed/handled 13 PRs this week, so I believe that we can handle one more :)

**Comment by GregOwen** [comment]
One sec; making sure that Sigma knows about the CLA and is comfortable with me signing it

**Comment by debugmiller** [comment]
Would love to see this merged in!

**Comment by GregOwen** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by GregOwen** [comment]
@sfc-gh-pfus I've rebased and we've signed the CLA (it took a while to get legal approval for that). Could you please kick off the tests?

File: statement_test.go
Comment (sfc-gh-pfus) [inline statement_test.go:843]: We have some assert functions, can you use them?
https://github.com/snowflakedb/gosnowflake/blob/master/assert_test.go

File: util.go
Comment (sfc-gh-pfus) [inline util.go:126]: Can you add some info to doc.go? I'd prefer having it before `Query request ID` section.

**Review by sfc-gh-pfus — CHANGES_REQUESTED** [review]

**Comment by sfc-gh-pfus** [comment]
Hi @GregOwen ! Thanks again for your contribution. I left some two suggestions and in the meantime started jobs.

**Comment by sfc-gh-pfus** [comment]
Your test hangs. I believe this is because you forgot to defer rows closing.

**Comment by GregOwen** [comment]
@sfc-gh-pfus I've pushed changes that should address your comments; let me know if they look good to you

**Review by sfc-gh-pfus — APPROVED** [review]

**Comment by sfc-gh-pfus** [comment]
Hi @GregOwen ! All of your checks are green (I fixed a flaky test in the meantime). Can you rebase your PR? I can't do it myself.

**Comment by GregOwen** [comment]
@sfc-gh-pfus rebased. It looks like we may get stuck in a "need to rebase" --> "need approval to run tests" loop since we're in different timezones. I'm based in EST; should we try to coordinate tomorrow so that I make my changes first thing in the morning my time (which I'm guessing will line up with the end of your workday)?

**Comment by sfc-gh-pfus** [comment]
Hi @GregOwen ! Actually I wanted to do it in the morning, before no one from merges from my team ;) Thank you very much, your commit will be released in January!

**Pull request merged**
